### PR TITLE
grpc-js: Add XdsClient class

### PR DESCRIPTION
**Pull request by murgatroid99**
This class will be used for all xDS gRPC operations. This PR just implements the EDS operations. The other operations will be implemented along with their corresponding load balancer implementations.

The interesting files here are `xds-client.ts` and `xds-bootstrap.ts`. The rest are submodules, generated code, and dependency updates.

### CODE REVIEW

**Comment by murgatroid99** [comment]
This is pretty much ready for review, but I want to keep it from being merged while it has a dependency on a prerelease version of `@grpc/proto-loader`. That can be fixed after #1474 is merged and published.

File: packages/grpc-js/src/xds-client.ts [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+  ) {
+    trace('Watcher removed for endpoint ' + edsServiceName);
+    const watchersEntry = this.endpointWatchers.get(edsServiceName);
+    let removedServiceName = false;
+    if (watchersEntry !== undefined) {
+      const entryIndex = watchersEntry.indexOf(watcher);
+      if (entryIndex >= 0) {
+        watchersEntry.splice(entryIndex, 1);
+      }
+      if (watchersEntry.length === 0) {
```
Comment (stanley-cheung) [inline packages/grpc-js/src/xds-client.ts:374]: Does comparing with 0 have to be `===`? I guess it doesn't hurt.

File: packages/grpc-js/src/xds-client.ts [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+            message.version_info,
+            message.nonce
+          );
+      }
+    });
+    this.adsCall.on('error', (error: ServiceError) => {
+      trace(
+        'ADS stream ended. code=' + error.code + ' details= ' + error.details
+      );
+      this.adsCall = null;
```
[the remark this answers is not in the document]
Comment (murgatroid99) [reply packages/grpc-js/src/xds-client.ts:224]: In short, yes.

One of the benefits of using arrow functions (`(args...) => {/*...*/}`) is that they are bound to the scope where they are declared, not where they are called. The `this` reference is valid in the `maybeStartAdsStream` method where this callback is declared, so it is valid here too.

File: packages/grpc-js/src/xds-client.ts [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+      this.reportStreamError(error);
+      /* Connection backoff is handled by the client object, so we can
+       * immediately start a new request to indicate that it should try to
+       * reconnect */
+      this.maybeStartAdsStream();
+    });
+    const endpointWatcherNames = Array.from(this.endpointWatchers.keys());
+    if (endpointWatcherNames.length > 0) {
+      this.adsCall.write({
```
[the remark this answers is not in the document]
Comment (murgatroid99) [reply packages/grpc-js/src/xds-client.ts:233]: One thing to note here is that `AggregatedDiscoveryService` is a service definition provided by Envoy that we are using. We don't decide anything about how the service is defined.

Also, the general `StreamAggregatedServices` workflow looks like this:

 1. The client a request message with a list of names of resources of a specific type. Later, we can send a new message with a different list of names to change what resources we are interested in.

 2. The server sends response message with a list of resources of that type. The server sends a new message whenever a resource is added, removed, or changed.

 3. For each response message, the client either acknowledges (`ack`) or rejects (`nack`) the update.

Within a single stream, each of these things happens independently for each type URL.

Hopefully this makes it clearer why we use a bidi stream for this method.

File: packages/grpc-js/src/xds-client.ts [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+  private reportStreamError(status: StatusObject) {
+    for (const watcherList of this.endpointWatchers.values()) {
+      for (const watcher of watcherList) {
+        watcher.onTransientError(status);
+      }
+    }
+    // Also do the same for other types of watchers when those are implemented
+  }
+
+  addEndpointWatcher(
```
[the remark this answers is not in the document]
Comment (murgatroid99) [reply packages/grpc-js/src/xds-client.ts:344]: Yes, `updateEdsNames` tells the server the new list of EDS names that we are interested in. It should be called whenever a name is added or removed from the list.

File: packages/grpc-js/src/xds-client.ts (hunk 1 above)
Comment (murgatroid99) [reply packages/grpc-js/src/xds-client.ts:374]: The convention is that all comparisons should be done using `===`.

**Pull request merged**